src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1993,62 +2038,118 @@ SubqueryPushdownMultiPlanTree(Query *queryTree)
 {
 	List *targetEntryList = queryTree->targetList;
 	List *qualifierList = NIL;
-	List *qualifierColumnList = NIL;
-	List *targetListColumnList = NIL;
 	List *columnList = NIL;
 	ListCell *columnCell = NULL;
 	MultiCollect *subqueryCollectNode = CitusMakeNode(MultiCollect);
 	MultiTable *subqueryNode = NULL;
-	MultiSelect *selectNode = NULL;
 	MultiProject *projectNode = NULL;
 	MultiExtendedOp *extendedOpNode = NULL;
 	MultiNode *currentTopNode = NULL;
 	RangeTblEntry *subqueryRangeTableEntry = NULL;
-	List *subqueryEntryList = SubqueryEntryList(queryTree);
-
+	Query *lowerQuery = NULL;
+	List *subqueryTargetEntryList = NIL;
+	List *columnNamesList = NIL;
+	StringInfo rteName = makeStringInfo();
+	List *addedColumnList = NIL;
+	List *havingClauseColumnList = NIL;
 
 	/* verify we can perform distributed planning on this query */
-	ErrorIfQueryNotSupported(queryTree);
-	ErrorIfSubqueryJoin(queryTree);
+	ErrorIfQueryNotSupported(queryTree, true);
 
 	/* extract qualifiers and verify we can plan for them */
 	qualifierList = QualifierList(queryTree->jointree);
 	ValidateClauseList(qualifierList);
 
 	/*
-	 * We disregard pulled subqueries. This changes order of range table list.
-	 * We do not allow subquery joins, so we will have only one range table
-	 * entry in range table list after dropping pulled subquery. For this reason,
-	 * here we are updating columns in the most outer query for where clause
-	 * list and target list accordingly.
+	 * We would be creating a new Query and pushing down top level query's
```

Comment:
more like
```
	 * Consider query
	 * 		select s1.a, sum(s2.c)
	 * 		from (some subquery) s1, (some subquery) s2
	 * 		where s1.a = s2.a
	 * 		group by s1.a
	 * 		having avg(s2.b);
	 *
	 * We want to prepare a multi tree to avoid subquery joins at top level,
	 * therefore above query is converted to an equivalent
	 * 	 	select worker_column_0, sum(worker_column_1)
	 * 	 	from (
	 * 	 	    	select s1.a as worker_column_0, s2.c as worker_column_1, s2.b as as worker_column_2
	 * 				from (some subquery) s1, (some subquery) s2
	 * 				where s1.a = s2.a
	 * 				) citus_subquery_name
	 * 		group by worker_column_0
	 * 		having avg(worker_column_2);
	 * 	After this conversion MultiTree is created as follows
	 * 	
	 * 	MultiExtendedOpNode(
	 * 		targetList : worker_column_0, sum(worker_column_1)
	 * 		groupBy : worker_column_0
	 * 		having :  avg(worker_column_2))
	 * --->MultiProject (worker_column_0, worker_column_1, worker_column_2)
	 * --->--->	MultiTable (subquery : citus_subquery_name)
	 * 
	 * Master and worker queries will be created out of this MultiTree at later stages.
```